### PR TITLE
[Feature] Progress bars

### LINKED ISSUE #21: [feature request] progress bars
We have the spinner which works well, but it would be awesome to also have a progressbar :smile:  Maybe an integration with `indicatif` that automatically sets indicatif styling based on the cliclack theme and sets the prefix characters so that the '│` on the left doesn't break? That would be awesome -- but realistically I'd be overjoyed with just a simple progress+download bar built-in and drop indicatif vs. having to hack up indicatif templates/styling to not break the cliclack look.

EDIT: Aha, and I just noticed that cliclack is already using indicatif.. so we're half way there! :smile: I'm pretty swamped atm, but if you guys don't have time in the near future to look at these I'd be happy to PR some of my feature requests when I get some downtime.

### PR DESCRIPTION
**Pull request by cylewitruk**
Closes #21 

These changes do what _I_ want them to do. I tried to (in all haste) follow how you guys have things set-up, so I hope it looks OK, but just let me know otherwise (or you're welcome to jump in and take over if it's just small nits if you want). :smile: 

Includes examples for both progress and download bars, including using `mpsc` to signal ctrl-c in the middle of an eventually async task.

Disclaimer: I didn't really read into how the `cliclack` theming works as I'm not currently using it (looks good otob :sweat_smile:) If I've missed or done something wrong there please let me know -- I really did mean "in all haste" above :see_no_evil: 

I can imagine wanting to support changing `progress_chars` etc. But otherwise I think you guys have something here with being slightly opinionated when it comes to formatting.

I'm not amazingly pleased with the handling of `<Enter>` here -- if you know how better to handle this that would be :pray:

### CODE REVIEW

**Comment by fadeevab** [comment]
@cylewitruk Thank you for the PR! Changes are gonna be requested. :)

From what I see, `src/downloadbar.rs` and `src/progressbar.rs` are heavily copy-pasted, the only difference is the format.
All those bars are `ProgressBars`.

1. Need to keep the `src/progress.rs` **only** with some default progress bar format for `start()` and expose the ability to set the `indicatif`'s format via `start_with_format(&mut self, message, format)`.
2. Style suggestion

❌ Not the `cliclack` style 
image
✔ Okay-ish style of the active prompt.
image

To be honest I would consider making it a one-liner (maybe even without a spinner as it already contains an animation):

```
│  
◐  [00:00:02] Installation ###########################>------------ (69/100)
```

File: src/progressbar.rs [hunk 1]
```diff
@@ -0,0 +1,77 @@
+use std::{fmt::Display, time::Duration};
+
+use indicatif::ProgressStyle;
+
+use crate::{theme::THEME, ThemeState};
+
+/// A spinner + progressbar that renders progress indication using current/total
+/// semantics. If you're looking for a download bar (or a bar that deals with
+/// bytes and formatting of bytes/kb/mb/gb, etc.), see [`DownloadBar`](crate::DownloadBar).
```
Comment (fadeevab) [inline src/progressbar.rs:9]: `Bytes/KB/MB/GB`

File: src/theme.rs [hunk 2]
```diff
@@ -456,6 +456,81 @@ pub trait Theme {
...
+    /// for the [`indicatif::ProgressBar`] progressbar behavior which disrupts
+    /// the line after the stop message reproduced while terminal resizing
+    /// (see [`Spinner::stop`](fn@crate::Spinner::stop)).
+    fn format_downloadbar_with_state(
+        &self,
+        msg: &str,
+        state: &ThemeState,
+    ) -> std::io::Result<String> {
+        let term = Term::stderr();
+        term.move_cursor_up(1)?;
```
Comment (fadeevab) [inline src/theme.rs:506]: This is not good, as `src/theme.rs` is de-coupled with any Terminal interfaces/abstractions. It must not be here.
You won't need it if you agree to implement a one-liner progress bar.

Comment (cylewitruk) [reply src/theme.rs:506]: I agree it's ugly -- but from wishing a "uniform UI experience" where progress bars are in alignment with each other (think the `indicatif:multi` example), I'm not sure how this should be handled while maintaining the **beautiful** look of `cliclack`. I'm admittedly only an hour or so deep in your codebase, so I'm hoping that you might have some bright ideas here? :smile:

**Comment by cylewitruk** [comment]
> Changes are gonna be requested. :)

Haha, what kind of maintainer would you be if you didn't :carpentry_saw: at PR's? :laughing: 

> From what I see, `src/downloadbar.rs` and `src/progressbar.rs` are heavily copy-pasted, the only difference is the format. All those bars are `ProgressBars`.

Yeah, they are basically copy-pase. I'm usually a DRY-advocate, but in this case I do see a value in keeping them separate as they may evolve separately -- the implementation I'm providing here is very rudimentary... I do envision wishes for differences between raw "1/32"-ish progress bars, bytes downloaded, etc. vs. other types. 

`indicatif` has solved the above with their templates, which are being used here ofc., but I think one of the biggest draws to `cliclack` (at least for me) is to **avoid** template syntax and "magically get" beautiful results using the framework's provided super-simple and opinionated impls, while still being able to adjust some colors etc.

I'm using `indicatif` together with `inquire` heavily today, and testing replacing both with `cliclack` (using this PR's changes anyway) are like __HEAVEN__ :smiley: :pray:

**Comment by cylewitruk** [comment]
> 1. Need to keep the `src/progress.rs` **only** with some default progress bar format for `start()` and expose the ability to set the `indicatif`'s format via `start_with_format(&mut self, message, format)`.

I agree with allowing flexibility, but one of the key reasons this project is attractive (to me!) is that it's **opinionated** _but not restrictively so_... i.e. not having to think about templates, and only dealing with a high-level theme. I'm currently heavily using `indicatif`+`inquire`, which are both very customizable... but they both use different color libraries and styling etc. so it's not an amazing experience.

File: src/theme.rs (hunk 2 above)
Comment (cylewitruk) [reply src/theme.rs]: Moved this to `progressbar.rs`-- is it OK there?

**Comment by cylewitruk** [comment]
I had a little extra time so I re-visited the implementation :smile: Please re-review.

> @cylewitruk Thank you for the PR! Changes are gonna be requested. :)
> 
> From what I see, `src/downloadbar.rs` and `src/progressbar.rs` are heavily copy-pasted, the only difference is the format. All those bars are `ProgressBars`.
> 
>     1. Need to keep the `src/progress.rs` **only** with some default progress bar format for `start()` and expose the ability to set the `indicatif`'s format via `start_with_format(&mut self, message, format)`.

Refactored this!

>     2. Style suggestion

Fixed!

> To be honest I would consider making it a one-liner (maybe even without a spinner as it already contains an animation):
> 
> ```
> │  
> ◐  [00:00:02] Installation ###########################>------------ (69/100)
> ```

The next thing on my list is multi-progress bars, so multi-line will be needed (for my case, anyway) in any case.

**Comment by fadeevab** [comment]
@cylewitruk This PR will be harder to review than the other ones. For instance, `###########################>` style is questionable from my perspective: it doesn't fit the style. However, someone would like it. Thus, we need to add customization. It brings complexity to the API and the code base.

IMO, it would make sense to tell users to use `indicatiff` directly mocking the `cliclack` than re-exporting the whole `indicatiff` API through the `cliclack`.

**Comment by cylewitruk** [comment]
> @cylewitruk This PR will be harder to review than the other ones. For instance, `###########################>` style is questionable from my perspective: it doesn't fit the style. However, someone would like it. Thus, we need to add customization. It brings complexity to the API and the code base.

I agree with regards to `####>` -- I just copied that styling from one of `indicatif`'s examples. We could easily add a `.with_progress_chars(...)` or so which allows the user to override, while finding a "nice" default style which you feel fits together with the other defaults. There's a whole bunch of ASCII progress bars out there that look nice, it's just to pick one :smile: 

> IMO, it would make sense to tell users to use `indicatiff` directly mocking the `cliclack` than re-exporting the whole `indicatiff` API through the `cliclack`.

Eh, I'm coming from `indicatif` + `inquire` ... agree with not re-exporting the entire `indicatif` API, but having it wrapped with sensible defaults which match the rest of the styling is a huge plus & time-saver imo. I might suggest wrapping everything "`indicatif`" in a feature though, so that it's not a forced dependency.

**Comment by cylewitruk** [comment]
@fadeevab Should we close this PR in favor of #27 since that's built on this anyway and has had more work on styling etc.?

**Comment by fadeevab** [comment]
> @fadeevab Should we close this PR in favor of #27 since that's built on this anyway and has had more work on styling etc.?

Yes

**Pull request closed without merging**